src/main/java/io/airlift/compress/zstd/ZstdNative.java

```diff
@@ -0,0 +1,184 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.compress.zstd;
+
+import io.airlift.compress.internal.NativeLoader;
+import io.airlift.compress.internal.NativeSignature;
+
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static java.lang.invoke.MethodHandles.lookup;
+
+final class ZstdNative
+{
+    private record MethodHandles(
+            @NativeSignature(name = "ZSTD_compressBound", returnType = long.class, argumentTypes = long.class)
+            MethodHandle maxCompressedLength,
+            @NativeSignature(name = "ZSTD_compress", returnType = long.class, argumentTypes = {MemorySegment.class, long.class, MemorySegment.class, long.class, int.class})
+            MethodHandle compress,
+            @NativeSignature(name = "ZSTD_decompress", returnType = long.class, argumentTypes = {MemorySegment.class, long.class, MemorySegment.class, long.class})
+            MethodHandle decompress,
+            @NativeSignature(name = "ZSTD_getFrameContentSize", returnType = long.class, argumentTypes = {MemorySegment.class, long.class})
+            MethodHandle uncompressedLength,
+            @NativeSignature(name = "ZSTD_isError", returnType = int.class, argumentTypes = long.class)
+            MethodHandle isError,
+            @NativeSignature(name = "ZSTD_getErrorName", returnType = MemorySegment.class, argumentTypes = long.class)
+            MethodHandle getErrorName,
+            @NativeSignature(name = "ZSTD_defaultCLevel", returnType = int.class, argumentTypes = {})
+            MethodHandle defaultCLevel) {}
+
+    private ZstdNative() {}
+
+    private static final Optional<LinkageError> LINKAGE_ERROR;
+    private static final MethodHandle MAX_COMPRESSED_LENGTH_METHOD;
+    private static final MethodHandle COMPRESS_METHOD;
+    private static final MethodHandle DECOMPRESS_METHOD;
+    private static final MethodHandle UNCOMPRESSED_LENGTH_METHOD;
+    private static final MethodHandle IS_ERROR_METHOD;
+    private static final MethodHandle GET_ERROR_NAME_METHOD;
+
+    // TODO should we just hardcode this to 3?
+    public static final int DEFAULT_COMPRESSION_LEVEL;
+
+    static {
+        NativeLoader.Symbols<MethodHandles> symbols = NativeLoader.loadSymbols("zstd", MethodHandles.class, lookup());
+        LINKAGE_ERROR = symbols.linkageError();
+        MethodHandles methodHandles = symbols.symbols();
+        MAX_COMPRESSED_LENGTH_METHOD = methodHandles.maxCompressedLength();
+        COMPRESS_METHOD = methodHandles.compress();
+        DECOMPRESS_METHOD = methodHandles.decompress();
+        UNCOMPRESSED_LENGTH_METHOD = methodHandles.uncompressedLength();
+        IS_ERROR_METHOD = methodHandles.isError();
+        GET_ERROR_NAME_METHOD = methodHandles.getErrorName();
+        try {
```

Comment:
this should be guarded with `if (LINKAGE_ERROR.isEmpty())`

so that we can call `ZstdNative.isEnabled()` to detect whether library loads properly. Right now if it doesn't it always fails and we cannot recover